alonzo/formal-spec/properties.tex

```diff
@@ -197,45 +197,210 @@ \section{Formal Properties}
 Only the $\mathsf{UTXOS}$ transition affects the ledger UTxO in Alonzo.
 
 \begin{lemma}
-  For all environments $e$, transitions $\mathsf{TRNS}$, transactions $tx$ and states $s, s'$ such that
+  For all environments $e$, transitions $\mathsf{TRNS}$, signals $f$ and states $s, s'$ such that
   \begin{equation*}
-    e\vdash s\trans{trns}{tx}s',
+    e\vdash s\trans{trns}{f}s',
   \end{equation*}
-  The following holds for some states $u, u'$ and environment $e'$ :
+  The following holds :
   \begin{equation*}
-    \Utxo(s) \neq \Utxo (s')~ \Rightarrow~e'\vdash u\trans{utxos}{tx}u'~\wedge~\Utxo(s)=\Utxo (u)~\wedge~\Utxo(s')=\Utxo (u')
+    \Utxo(s) \neq \Utxo (s')~ \Rightarrow~\mathsf{TRNS}~=~\mathsf{UTXOS}
   \end{equation*}
 \end{lemma}
 \begin{proof}
-  The $\mathsf{UTXOS}$ transition is the only one that updates the UTxO, but not via
-  the result of another transition, so any update from the UTxO of $s$ to that of $s'$
+  By inspecting each transition in this specification, as well as those inherited from the
+  Shelley one, we see that any update from the UTxO of $s$ to that of $s'$
   must be done by $\mathsf{UTXOS}$.
 \end{proof}
 \end{property}
 
-\begin{enumerate}
-\item
-  \emph{Transaction Validation.}
-  If a Shelley transaction is accepted, it is fully processed.
+\begin{property}[Script interpreter arguments are fixed (deterministic script evaluation)]
+  \label{prop:fixed-inputs}
+
+For this property to hold, we make an important assumption about the Plutus script
+interpreter : it is a pure function that receives only the arguments provided by the ledger when it is
+invoked by calling the $\fun{runPLCScript}$ function. In particular, it does not
+obtain any system randomness, etc. In addition, we assume that the consensus
+function $\fun{epochInfoSlotToUTCTime}$ for converting a slot interval into a
+system time interval is deterministic. Furthermore, we assume that the two global
+constants, $\EpochInfo$ and $\SystemStart$, which it also takes as parameters,
+cannot change within the span of an interval on which $\fun{epochInfoSlotToUTCTime}$
+succeeds in doing the conversion.
+
+With this in mind, we can make the claim that
+all the data that a non-native script interpreter gets as input
+is fixed either by
+
+\begin{itemize}
+  \item the input itself (or its hash) being inside the transaction body --- and therefore signed, or
+  \item fixed by it (or its hash) having been recorded on the ledger
+\end{itemize}
+
+The fixed argument values provided by the ledger, together with the assumption that there
+are no additional data the interpreter can inspect, allows us to say that
+script evaluation is entirely deterministic. More precisely, we say that
+
+\begin{lemma}
+  For all environments $e, e'$, transactions $tx, tx'$ and states $s, s', u, u'$ such that
```

Comment:
I think it makes sense the way it is - we want to show that no matter to what UTxO the two transaction are applied to, so long as they are both phase-1 valid (there might be some differences in the UTxO entries the transaction is not spending), and the bodies of them are the same (ie the parts that the collateral key owners sign) , the validation outcomes must be the same

Reply:
Oh, I missed the condition that the bodies should be equal. Depending on how difficult we want the proof to be, requiring the utxos to be equal may still be necessary though. Otherwise, you could have two different utxos that just differ by one datum hash in such a way that changes the transaction outcome. To reason around that, we'd have to require both utxos to be some state that was reached using a sequence of transactions and use the "injectivity" of hash functions. An intermediate case would be to assume that the utxos agree on the part that is mentioned by the transaction, which is probably good enough for what we want and also a strong enough condition to prove the property.

Reply:
different datum associated with an TxIn would result in the TxIn itself (the hash of the tx containing the output with that datum) must also be different

Reply:
Only if you consider utxos that are obtained via transitions from a genesis block. Which then pulls in a bunch of additional complexity which doesn't seem worth it

Reply:
all you need is that in any UTxO entry, different TxIn => different TxOut

Reply:
Yes, but we'd need that for all utxos, which is false. If we're just talking about any two utxos, we can have `utxo1 = { in |-> out1 , ... }` and `utxo2 = { in |-> out2 , ...}` with `out1 /= out2`. We only get this property once we talk about utxos we obtained via application of rules.

Reply:
you're right, I will add that the states must be valid ledger states